src/content_mapblock.cpp

```diff
@@ -525,1633 +380,949 @@ static TileSpec getSpecialTile(const ContentFeatures &f,
-					tileindex = 1; // curved
-				if (is_rail_x_all[0] && is_rail_x_all[1])
-					angle = 90;
-				if (is_rail_z_all[0] && is_rail_z_all[1]) {
-					if (is_rail_z[4])
-						angle = 180;
-				}
-				else if (is_rail_x_all[0] && is_rail_z_all[0])
-					angle = 270;
-				else if (is_rail_x_all[0] && is_rail_z_all[1])
-					angle = 180;
-				else if (is_rail_x_all[1] && is_rail_z_all[1])
-					angle = 90;
-				break;
-			case 3:
-				// here is where the potential to 'switch' a junction is, but not implemented at present
-				tileindex = 2; // t-junction
-				if(!is_rail_x_all[1])
-					angle = 180;
-				if(!is_rail_z_all[0])
-					angle = 90;
-				if(!is_rail_z_all[1])
-					angle = 270;
-				break;
-			case 4:
-				tileindex = 3; // crossing
-				break;
-			case 5: //sloped
-				if (is_rail_z[4])
-					angle = 180;
-				if (is_rail_x[4])
-					angle = 90;
-				if (is_rail_x[5])
-					angle = -90;
-				break;
-			default:
-				break;
-			}
+void MapblockMeshGenerator::drawFirelikeQuad(float rotation, float opening_angle,
+	float offset_h, float offset_v)
+{
+	v3f vertices[4] = {
+		v3f(-scale, -BS/2, 0),
+		v3f( scale, -BS/2, 0),
+		v3f( scale, -BS/2 + scale * 2, 0),
+		v3f(-scale, -BS/2 + scale * 2, 0),
+	};
+	for (int i = 0; i < 4; i++) {
+		vertices[i].rotateYZBy(opening_angle);
+		vertices[i].Z += offset_h;
+		vertices[i].rotateXZBy(rotation);
+		vertices[i].Y += offset_v;
+	}
+	drawQuad(vertices);
+}
 
-			TileSpec tile = getNodeTileN(n, p, tileindex, data);
-			tile.material_flags &= ~MATERIAL_FLAG_BACKFACE_CULLING;
-			tile.material_flags |= MATERIAL_FLAG_CRACK_OVERLAY;
+void MapblockMeshGenerator::drawFirelikeNode()
+{
+	useTile(0, false);
+	scale = BS / 2 * f->visual_scale;
+
+	// Check for adjacent nodes
+	bool neighbors = false;
+	bool neighbor[6] = { 0, 0, 0, 0, 0, 0 };
+	content_t current = n.getContent();
+	for (int i = 0; i < 6; i++) {
+		v3s16 n2p = blockpos_nodes + p + g_6dirs[i];
+		MapNode n2 = data->m_vmanip.getNodeNoEx(n2p);
+		content_t n2c = n2.getContent();
+		if (n2c != CONTENT_IGNORE && n2c != CONTENT_AIR && n2c != current) {
+			neighbor[i] = true;
+			neighbors = true;
+		}
+	}
+	bool drawBasicFire = neighbor[D6D_YN] || !neighbors;
+	bool drawBottomFire = neighbor[D6D_YP];
+
+	if (drawBasicFire || neighbor[D6D_ZP])
+		drawFirelikeQuad(  0, -10, 0.4 * BS);
+	else if (drawBottomFire)
+		drawFirelikeQuad(  0, 70, 0.47 * BS, 0.484 * BS);
+
+	if (drawBasicFire || neighbor[D6D_XN])
+		drawFirelikeQuad( 90, -10, 0.4 * BS);
+	else if (drawBottomFire)
+		drawFirelikeQuad( 90, 70, 0.47 * BS, 0.484 * BS);
+
+	if (drawBasicFire || neighbor[D6D_ZN])
+		drawFirelikeQuad(180, -10, 0.4 * BS);
+	else if (drawBottomFire)
+		drawFirelikeQuad(180, 70, 0.47 * BS, 0.484 * BS);
+
+	if (drawBasicFire || neighbor[D6D_XP])
+		drawFirelikeQuad(270, -10, 0.4 * BS);
+	else if (drawBottomFire)
+		drawFirelikeQuad(270, 70, 0.47 * BS, 0.484 * BS);
+
+	if (drawBasicFire) {
+		drawFirelikeQuad(45, 0, 0.0);
+		drawFirelikeQuad(-45, 0, 0.0);
```

Comment:
Line 977 to here: The alignment of number right edges looks good but is inconsistent.